File: jvm-experiments-commons/src/main/java/kr/kth/commons/timeperiod/timerange/TimeRange.java
package kr.kth.commons.timeperiod.timerange;

import kr.kth.commons.timeperiod.ITimePeriod;
import kr.kth.commons.timeperiod.ITimeRange;
import kr.kth.commons.timeperiod.TimePeriodBase;
import org.joda.time.DateTime;

/**
 * Time 기준으로 기간을 표현합니다. 단 하한(시작시각) 과 상한(완료시각) 은 Open 일 수 있습니다. 즉 개구간일 수 있습니다.
 * User: sunghyouk.bae@gmail.com
 * Date: 12. 9. 18.
 */
public class TimeRange extends TimePeriodBase implements ITimeRange {

	private static final long serialVersionUID = -8743222521934888987L;

	public static final TimeRange Anytime = new TimeRange(true);

	// region << Constructors >>

	public TimeRange() {
		super();
	}

	public TimeRange(boolean readonly) {
		super(readonly);
	}

	public TimeRange(DateTime moment) {
		super(moment, false);
	}

	public TimeRange(DateTime moment, boolean readonly) {
		super(moment, readonly);
	}

	public TimeRange(DateTime start, DateTime end) {
		super(start, end, false);
	}

	public TimeRange(DateTime start, DateTime end, boolean readonly) {
		super(start, end, readonly);
	}

	public TimeRange(DateTime start, long duration) {
		super(start, duration, false);
	}

	public TimeRange(DateTime start, long duration, boolean readonly) {
		super(start, duration, readonly);
	}

	public TimeRange(ITimePeriod src) {
		super(src);
	}

	public TimeRange(ITimePeriod src, boolean readonly) {
		super(src, readonly);
	}

	// endregion

	@Override
	public void setStart(DateTime start) {
		assertMutable();
		this.start = start;
	}

	@Override
	public void setEnd(DateTime end) {
		assertMutable();
		this.end = end;
	}

	@Override
	public void setDuration(long duration) {
		assertMutable();
		this.end = getStart().plus(duration);
	}

	@Override
	public void expandStartTo(DateTime moment) {
		assertMutable();
		this.start = moment;
	}

	@Override
	public void expandEndTo(DateTime moment) {
		assertMutable();
		this.end = moment;
	}

	@Override
	public void expandTo(DateTime moment) {
		assertMutable();

		if (moment.compareTo(start) <= 0)
			setStart(moment);
		else if (moment.compareTo(end) >= 0)
			setEnd(moment);
	}

	@Override
	public void expandTo(ITimePeriod period) {
		assertMutable();
		expandStartTo(period.getStart());
		expandEndTo(period.getEnd());
	}

	@Override
	public void shrinkStartTo(DateTime moment) {
		assertMutable();
		this.start = moment;
	}

	@Override
	public void shrinkEndTo(DateTime moment) {
		assertMutable();
		this.end = moment;
	}

	@Override
	public void shrinkTo(ITimePeriod period) {
		assertMutable();
		shrinkStartTo(period.getStart());
		shrinkEndTo(period.getEnd());
	}
}

File: jvm-experiments-timeperiod/src/main/scala/kr/kth/timeperiod/TimeRange.scala
package kr.kth.timeperiod

import kr.kth.commons.Guard
import kr.kth.commons.slf4j.Logging
import org.joda.time.DateTime

/**
 * 시각 구간 ( 시작 ~ 완료 시각 )을 나타냅니다.
 * User: sunghyouk.bae@gmail.com
 * Date: 13. 1. 14
 */
class TimeRange(start: DateTime, end: DateTime, readonly: Boolean)
    extends TimePeriod(start, end, readonly) with ITimeRange {

    def this(source: ITimePeriod, readonly: Boolean) {
        this(source.getStart, source.getEnd, readonly)
    }

    def this(source: ITimePeriod) {
        this(source, false)
    }

    def this(moment: DateTime, readonly: Boolean) {
        this(moment, moment, readonly)
    }

    def this(start: DateTime, end: DateTime) {
        this(start, end, false)
    }

    def this() {
        this(TimeSpec.MinPeriodTime, TimeSpec.MaxPeriodTime, false)
    }

}

/**
 * TimeRange Object
 */
object TimeRange extends Logging {

    lazy val Anytime: TimeRange = apply(readonly = true)

    def apply(start: DateTime = TimeSpec.MinPeriodTime,
              end: DateTime = TimeSpec.MaxPeriodTime,
              readonly: Boolean = false): TimeRange =
        new TimeRange(start, end, readonly)

    def apply(source: ITimePeriod): TimeRange = {
        Guard.shouldNotBeNull(source, "source")
        if (source == Anytime) Anytime
        else apply(source.getStart, source.getEnd, source.isReadonly)
    }

    def apply(source: ITimePeriod, readonly: Boolean): TimeRange = {
        Guard.shouldNotBeNull(source, "source")
        apply(source.getStart, source.getEnd, readonly)
    }

    def unapply(range: ITimeRange) = (range.getStart, range.getEnd, range.isReadonly)
}

/**
 * ITimeRange trait
 */
trait ITimeRange extends ITimePeriod {

    /**
     * 시작시각을 설정합니다.
     */
    def setStart(newStart: DateTime) {
        assertMutable()
        _start = newStart
    }

    /**
     * 완료시각을 설정합니다.
     */
    def setEnd(newEnd: DateTime) {
        assertMutable()
        _end = newEnd
    }

    /**
     * 시작시각을 지정된 시각으로 변경합니다. 기존 시작시각보다 작은 값 (과거) 이여야 합니다.
     */
    def expandStartTo(moment: DateTime) {
        assertMutable()

        if (_start.compareTo(moment) > 0) {
            log.debug("expand start[{}] to [{}]", _start, moment)
            _start = moment
        }
    }

    /**
     * 완료시각을 지정된 시각으로 변경합니다. 기존 완료시각보다 큰 값 (미래) 이어야 합니다.
     */
    def expandEndTo(moment: DateTime) {
        assertMutable()

        if (_end.compareTo(moment) < 0) {
            log.debug("expand end[{}] to [{}]", _end, moment)
            _end = moment
        }
    }

    /**
     * 시작시각과 완료시각을 지정된 시각으로 설정합니다.
     */
    def expandTo(moment: DateTime) {
        assertMutable()
        expandStartTo(moment)
        expandEndTo(moment)
    }

    /**
     * 시작시각과 완료시각을 지정된 기간 정보를 기준으로 변경합니다.
     */
    def expandTo(period: ITimePeriod) {
        assertMutable()
        Guard.shouldNotBeNull(period, "period")
        if (period.hasStart) expandStartTo(period.getStart)
        if (period.hasEnd) expandEndTo(period.getEnd)
    }

    /**
     * 시작시각을 지정된 시각으로 변경합합니다. 기존 시작시각보다 큰 값(미래) 이어야 합니다.
     */
    def shrinkStartTo(moment: DateTime) {
        assertMutable()
        if (_start.compareTo(moment) < 0) {
            log.debug("shrink start=[{}] to [{}]", _start, moment)
            _start = moment
        }
    }

    /**
     * 완료시각을 지정된 시각으로 변경합니다. 기존 완료시각보다 작은 값(과거) 이어야 하고, 시작시각보다는 같거나 커야 합니다.
     */
    def shrinkEndTo(moment: DateTime) {
        assertMutable()
        if (_start.compareTo(moment) < 0) {
            log.debug("shrink end=[{}] to [{}]", _end, moment)
            _end = moment
        }
    }

    /**
     * 기간을 지정된 기간으로 축소시킵니다.
     */
    def shrinkTo(period: ITimePeriod) {
        assertMutable()
        Guard.shouldNotBeNull(period, "period")

        if (period.hasStart) shrinkStartTo(period.getStart)
        if (period.hasEnd) shrinkEndTo(period.getEnd)
    }
}
